Optimizes fill_halo_regions_open.jl

jagoosw opened the pull request:
I was doing some profiling on a model with no open boundaries and discovered that this function was causing a big slow down. I guess this is because the compiler isn't managing to work out its just a load of nothing operations but this change appears to make it completely go away.

simone-silvestri commented:
what was the slowdown?

glwagner commented:
Might be good to profile with `fill_open_boundary_regions!(fields::NTuple, boundary_conditions, indices, loc, grid, args...; kwargs...) = nothing` also to check this

glwagner commented on src/BoundaryConditions/fill_halo_regions_open.jl:
```diff
@@ -22,8 +22,9 @@ function fill_open_boundary_regions!(field, boundary_conditions, indices, loc, g
-fill_open_boundary_regions!(fields::NTuple, boundary_conditions, indices, loc, grid, args...; kwargs...) =
-    [fill_open_boundary_regions!(field, boundary_conditions[n], indices, loc[n], grid, args...; kwargs...) for (n, field) in enumerate(fields)]
+
+fill_open_boundary_regions!(fields::NTuple, boundary_conditions, indices, loc, grid, args...; kwargs...) = 
+    ntuple(n->fill_open_boundary_regions!(fields[n], boundary_conditions[n], indices, loc[n], grid, args...; kwargs...), Val(length(fields)))
```
```suggestion
function fill_open_boundary_regions!(fields::NTuple, bcs, indices, loc, args...; kw...) 
    N = length(fields)
    return ntuple(n -> fill_open_boundary_regions!(fields[n], bcs[n], indices, loc[n], args...; kw...), Val(N))
end
```

glwagner replied:
If we can't read the code, we've little to show for all our hard work

glwagner commented:
The profile should be documented here. I'll tentatively approach but the documentation is important

jagoosw commented:
Ah so I've realised this isn't the fix we needed, and I was just hiding it from myself in the profile because I replaced the function by writing it in the REPL. I made an MWE:

```jula
using Oceananigans

grid = RectilinearGrid(GPU(), topology = (Flat, Flat, Bounded), size = (100, ), extent = (400, ))

model = HydrostaticFreeSurfaceModel(; grid, velocities = PrescribedVelocityFields(), momentum_advection=nothing, buoyancy=nothing, tracers = ntuple(n->Symbol(:T, n), Val(30)))
```
<img width="1361" alt="Screenshot 2024-09-26 at 12 00 29" src="https://github.com/user-attachments/assets/2cf2a379-88e9-428f-8156-4ddead2a02e4">
You can see from this profile that `fill_open_boundary_regions!` takes a lot longer than `fill_halo_event!`, even though there are no velocity open boundaries. This is because it is launching a load of zero size kernels where as `fill_halo_event!` just returns nothing instead.

I've fixed this now and get this from the profile instead:
<img width="1251" alt="Screenshot 2024-09-26 at 12 02 11" src="https://github.com/user-attachments/assets/0bf91086-bc12-4a17-ba48-89b9b1c2e7ae">

In numbers, the original version benchmarks `time_step!` at around 4.074 ms ± 581.472 μs and the new version 2.438 ms ± 501.642 μs

glwagner commented on src/BoundaryConditions/fill_halo_regions_open.jl:
```diff
@@ -17,13 +17,19 @@ function fill_open_boundary_regions!(field, boundary_conditions, indices, loc, g
 
-fill_open_boundary_regions!(fields::NTuple, boundary_conditions, indices, loc, grid, args...; kwargs...) =
-    [fill_open_boundary_regions!(field, boundary_conditions[n], indices, loc[n], grid, args...; kwargs...) for (n, field) in enumerate(fields)]
+@inline fill_open_halo_event!(open_fill, field, left_bc, right_bc, fill_size, loc, arch, grid, args) =
+    launch!(arch, grid, fill_size, open_fill, field, left_bc, right_bc, loc, grid, args)
```
```suggestion
@inline fill_open_halo_event!(fill_open_halo, field, left_bc, right_bc, fill_size, loc, arch, grid, args) =
    launch!(arch, grid, fill_size, fill_open_halo, field, left_bc, right_bc, loc, grid, args)
```

glwagner commented on src/BoundaryConditions/fill_halo_regions_open.jl:
```diff
@@ -17,13 +17,19 @@ function fill_open_boundary_regions!(field, boundary_conditions, indices, loc, g
     open_fill, regular_fill = get_open_halo_filling_functions(loc) 
```
This is a little convoluted. Why do you need a function to determine the size? Why can't we just use `loc`?

jagoosw replied:
This comes from normal halo filling and I think its main purpose is for windowed fields:
https://github.com/CliMA/Oceananigans.jl/blob/4dbec269c7fa88a611121a78924589945b535d2d/src/BoundaryConditions/fill_halo_regions.jl#L349-L433

glwagner commented on src/BoundaryConditions/fill_halo_regions_open.jl:
```diff
@@ -17,13 +17,19 @@ function fill_open_boundary_regions!(field, boundary_conditions, indices, loc, g
+
+@inline fill_open_boundary_regions!(fields::NTuple{N}, boundary_conditions, indices, loc, grid, args...; kwargs...) where N =
+    ntuple(n->fill_open_boundary_regions!(fields[n], boundary_conditions[n], indices, loc[n], grid, args...; kwargs...), 
+           Val(N))
```
more lines

jagoosw commented:
I'll stream line those two line before I merge

simone-silvestri commented:
Is this PR close to merging?

tomchor commented:
cc @ali-ramadhan

simone-silvestri commented:
I think this has been superseded by #3834. @jagoosw let us know if it is the case or not.

jagoosw commented:
> I think this has been superseded by #3834. @jagoosw let us know if it is the case or not.

Yeah definitely superseded!
